x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/profile/RestSearchProfilesAction.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.rest.action.profile;
+
+import org.elasticsearch.client.internal.node.NodeClient;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xpack.core.security.action.profile.SearchProfilesAction;
+import org.elasticsearch.xpack.core.security.action.profile.SearchProfilesRequest;
+import org.elasticsearch.xpack.security.rest.action.SecurityBaseRestHandler;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class RestSearchProfilesAction extends SecurityBaseRestHandler {
+
+    static final ConstructingObjectParser<Payload, Void> PARSER = new ConstructingObjectParser<>(
+        "search_profile_request_payload",
+        a -> new Payload((String) a[0], (Integer) a[1])
+    );
+
+    static {
+        PARSER.declareString(optionalConstructorArg(), new ParseField("query"));
+        PARSER.declareInt(optionalConstructorArg(), new ParseField("size"));
```

Comment:
> For the data parameter. I think we can easily support including and excluding fields. Do you think supporting excluding fields is a good idea?

The `data` parameter underlying uses the same code for `filter_path`. So excluding can be supported easily. But I suggest we also forbid any key to begin with dash (`-`) when we add the exclusion support. I can follow up with a separate PR if this sounds good to you.

> I think it's un-intuitive that some request parameters are in the body and others in the URL.
> I would suggest that all the parameters, data, query, size be URL parameters.
I agree in general that splitting parameters between body and URL is not great. But for this API, I think request body is unavoidable for more involved searches. So I'd put everything into request body instead of having them supported in URL. That said, the `data` parameter is based on `filter_path` and takes comma separated strings, which fits better as request parameter. It's likely the only parameter that stays in the URL and I personally think it is fine.

> I'm not a fan of the query name for the parameter.
Agreed. I don't like it either for the same reason you stated. I took it from the design doc and didn't change because (1) naming is hard and I couldn't  think of a better name; (2) if we give it a distinct name, it also implies we are willing to (have to?) provide different fields for searching different sections of the profile doc. This may be feasible or even disirable. But I wasn't sure at the time. If we are to provide different fields for searching different sections, there is also the question about how specific the name should be, e.g. `name` vs `user`? If we choose `user`, does that imply it can potentially contain other future user information that is not name related? On the other hand, if we choose `name`, the field can never be expanded for other usage. Between the two, I lean towards the later and changed the field to be `name`.

Reply:
> I agree in general that splitting parameters between body and URL is not great. But for this API, I think request body is unavoidable for more involved searches. So I'd put everything into request body instead of having them supported in URL. That said, the data parameter is based on filter_path and takes comma separated strings, which fits better as request parameter. It's likely the only parameter that stays in the URL and I personally think it is fine.

Begrudgingly OK... I would still prefer all parameters be in the request body, and maybe make the endpoint POST only.

Reply:
> What I was thinking was to use a different list for exclude, ala https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-fields.html#source-filtering .

Honestly I didn't know this exists. I like it. If we adopt this approach, I can see how everything can be kept in request body. But then the question is what we should do about GetProfileByID? It currently does not support a request body and filter data by request parameter. Are we ok with the two APIs having different approaches or should we make them consistent with each other? We should make this a separate issue to discuss about. I'll add a point to the task list.